File: internal/models/lead_sync.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// LeadSyncStatus is the lifecycle state of a saved lead-sync source. It is
// not a queue state — syncs are on-demand — only a record of how the last
// "Sync now" went.
type LeadSyncStatus string

const (
	// LeadSyncStatusIdle — never synced, or the last sync succeeded.
	LeadSyncStatusIdle LeadSyncStatus = "idle"
	// LeadSyncStatusSyncing — a sync is in flight (set for the duration of
	// the synchronous SyncNow call).
	LeadSyncStatusSyncing LeadSyncStatus = "syncing"
	// LeadSyncStatusError — the last sync failed; LastError carries why.
	LeadSyncStatusError LeadSyncStatus = "error"
)

// LeadSyncSource is a saved, re-runnable binding between a Google Sheet (read
// through an existing google_sheets OAuth connection) and Warmbly's contact
// importer. New rows create contacts; rows matching an existing contact by
// email are updated. A source optionally enrols new/updated leads into a
// campaign and/or tags them with categories.
//
// Secrets never live here — the sheet is read using the linked connection's
// envelope-encrypted Google token, resolved at sync time.
type LeadSyncSource struct {
	ID              uuid.UUID `json:"id"`
	OrganizationID  uuid.UUID `json:"organization_id"`
	CreatedByUserID uuid.UUID `json:"created_by_user_id"`
	Provider        string    `json:"provider"`
	ConnectionID    uuid.UUID `json:"connection_id"`
	SheetID         string    `json:"sheet_id"`
	SheetTitle      string    `json:"sheet_title,omitempty"`
	TabTitle        string    `json:"tab_title,omitempty"`
	A1Range         string    `json:"a1_range,omitempty"`
	HasHeader       bool      `json:"has_header"`

	// ColumnMapping reuses the exact contact-import mapping shape so the sheet
	// rows flow through the same /contacts/import/commit code path.
	ColumnMapping []ContactImportColumnMapping `json:"column_mapping"`
	Dedup         ContactImportDedupStrategy   `json:"dedup"`

	// TargetCampaignID, when set, enrols every new/updated lead into that
	// campaign on each sync.
	TargetCampaignID  *uuid.UUID `json:"target_campaign_id,omitempty"`
	CategoryIDs       []string   `json:"category_ids"`
	SubscribedDefault bool       `json:"subscribed_default"`

	Label  string         `json:"label,omitempty"`
	Status LeadSyncStatus `json:"status"`

	LastSyncedAt *time.Time           `json:"last_synced_at,omitempty"`
	LastResult   *ContactImportResult `json:"last_result,omitempty"`
	LastError    string               `json:"last_error,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// CreateLeadSyncSource is the input for creating a saved source. The connection
// must be the org's google_sheets OAuth connection.
type CreateLeadSyncSource struct {
	ConnectionID      uuid.UUID                    `json:"connection_id"`
	SheetID           string                       `json:"sheet_id"`
	SheetTitle        string                       `json:"sheet_title"`
	TabTitle          string                       `json:"tab_title"`
	HasHeader         bool                         `json:"has_header"`
	ColumnMapping     []ContactImportColumnMapping `json:"column_mapping"`
	Dedup             ContactImportDedupStrategy   `json:"dedup"`
	TargetCampaignID  *uuid.UUID                   `json:"target_campaign_id"`
	CategoryIDs       []string                     `json:"category_ids"`
	SubscribedDefault *bool                        `json:"subscribed_default"`
	Label             string                       `json:"label"`
}

// UpdateLeadSyncSource is the input for editing a saved source. All fields are
// optional pointers so a PATCH can touch a single setting; nil leaves the
// stored value untouched.
type UpdateLeadSyncSource struct {
	SheetID           *string                       `json:"sheet_id"`
	SheetTitle        *string                       `json:"sheet_title"`
	TabTitle          *string                       `json:"tab_title"`
	HasHeader         *bool                         `json:"has_header"`
	ColumnMapping     *[]ContactImportColumnMapping `json:"column_mapping"`
	Dedup             *ContactImportDedupStrategy   `json:"dedup"`
	TargetCampaignID  *uuid.UUID                    `json:"target_campaign_id"`
	ClearCampaign     bool                          `json:"clear_campaign"`
	CategoryIDs       *[]string                     `json:"category_ids"`
	SubscribedDefault *bool                         `json:"subscribed_default"`
	Label             *string                       `json:"label"`
}

// LeadSyncResult is what "Sync now" returns: the underlying contact-import
// counts plus the source id so the dashboard can attribute the run.
type LeadSyncResult struct {
	SourceID uuid.UUID            `json:"source_id"`
	Result   *ContactImportResult `json:"result"`
}

File: web/src/lib/api/models/app/leadsync/LeadSync.ts
// Mirror of the backend's models/lead_sync.go shapes. Field names match the
// JSON contract exactly. Lead-sync is an on-demand Google Sheet → contacts
// importer: a saved "sync source" the user re-runs via a "Sync now" button.
//
// We deliberately REUSE the contact-import column-mapping + result types from
// the import wizard rather than redeclaring them, so the same TargetPicker /
// MapStep mapper and result rendering work verbatim against /google/preview
// and /sources/:id/sync responses.

import type {
    ImportColumnMapping,
    ImportDedupStrategy,
    ImportPreview,
    ImportResult,
} from "@/lib/api/client/app/contacts/importContacts";

// Re-export the reused contact-import types under lead-sync names so callers
// can import everything lead-sync from one module.
export type {
    ImportColumnMapping,
    ImportDedupStrategy,
    ImportPreview,
    ImportResult,
};

// LeadSyncStatus mirrors models.LeadSyncStatus. The synchronous SyncNow only
// ever lands on idle/error; "syncing" exists for completeness/future async use.
export type LeadSyncStatus = "idle" | "syncing" | "error";

// A saved on-demand Google-Sheet → contacts source.
export interface LeadSyncSource {
    id: string;
    organization_id: string;
    created_by_user_id: string;
    provider: string;
    connection_id: string;
    sheet_id: string;
    sheet_title?: string;
    tab_title?: string;
    a1_range?: string;
    has_header: boolean;
    column_mapping: ImportColumnMapping[];
    dedup: ImportDedupStrategy;
    target_campaign_id?: string;
    category_ids: string[];
    subscribed_default: boolean;
    label?: string;
    status: LeadSyncStatus;
    last_synced_at?: string;
    last_result?: ImportResult;
    last_error?: string;
    created_at: string;
    updated_at: string;
}

// Result of POST /sources/:id/sync.
export interface LeadSyncResult {
    source_id: string;
    result: ImportResult;
}

// Reports whether the org has a connected (hidden) google_sheets OAuth
// connection usable for lead-sync.
export interface LeadSyncConnection {
    connected: boolean;
    connection: {
        id: string;
        external_account_name: string;
        status: string;
    } | null;
}

// One tab of a spreadsheet, from /google/spreadsheet.
export interface SheetTab {
    title: string;
    index: number;
}

// Spreadsheet metadata returned by /google/spreadsheet — drives the tab picker.
export interface SheetMeta {
    sheet_id: string;
    title: string;
    tabs: SheetTab[];
}

// Request body for creating a new source.
export interface CreateLeadSyncSource {
    connection_id: string;
    sheet_id: string;
    sheet_title: string;
    tab_title: string;
    has_header: boolean;
    column_mapping: ImportColumnMapping[];
    dedup: ImportDedupStrategy;
    target_campaign_id?: string;
    category_ids: string[];
    subscribed_default?: boolean;
    label: string;
}

// Request body for PATCH — all fields optional; omitted ⇒ unchanged.
// `clear_campaign: true` detaches the target campaign (a nil pointer alone
// can't express "clear").
export interface UpdateLeadSyncSource {
    sheet_id?: string;
    sheet_title?: string;
    tab_title?: string;
    has_header?: boolean;
    column_mapping?: ImportColumnMapping[];
    dedup?: ImportDedupStrategy;
    target_campaign_id?: string;
    clear_campaign?: boolean;
    category_ids?: string[];
    subscribed_default?: boolean;
    label?: string;
}
